lib/elixir/lib/exception.ex

```diff
@@ -2131,6 +2131,57 @@ defmodule UnicodeConversionError do
   end
 end
 
+defmodule MissingApplicationsError do
+  @moduledoc """
+  An exception that is raised when an application depends on one or more
+  missing applications.
+
+  This exception is used by Mix and other tools. It can also be used by library authors
+  when their library only requires an external application (like a dependency) for a subset
+  of features.
+
+  The fields of this exception are public. See `t:t/0`.
+
+  *Available since v1.18.0.*
+
+  ## Examples
+
+      unless Application.spec(:plug, :vsn) do
+        raise MissingApplicationsError,
+          description: "application :plug is required for testing Plug-related functionality",
+          apps: [{:plug, "~> 1.0"}]
+      end
+
+  """
+
+  @moduledoc since: "1.18.0"
+
+  @typedoc since: "1.18.0"
+  @type t() :: %__MODULE__{
+          apps: [{Application.app(), Version.requirement()}, ...],
```

Comment:
I mentioned earlier that a potential use case for this is raising on missing Elixir/OTP apps. If so, apps being non-empty makes it a bit awkward imo: 

```elixir
raise MissingApplicationsError,
        description: "please add :ssl to :extra_applications",
        apps: [{:ssl, ">= 0.0.0"}]
```

```
** (MissingApplicationsError) please add :ssl to :extra_applications

To address this, include these applications as your dependencies:

  {:ssl, ">= 0.0.0"}
```

It's OK if this exception is not intended for that. Alternatively, allow :apps to be empty?

Reply:
We can detect if the application comes from OTP and change the message accordingly. So no concerns here.